sdks/python/apache_beam/transforms/core.py

```diff
@@ -1465,6 +1465,11 @@ def get_restriction_coder(self):
     from apache_beam.runners.common import DoFnSignature
     return DoFnSignature(self.fn).get_restriction_coder()
 
+  def _add_type_constraint_from_consumer(self, full_label, input_type_hints):
```

Comment:
You're recording that the type hint comes from this transform, but are you recording that it comes from the input of this transform? (E.g. if I get a failure about transform B's type hint being violated, it may not be clear if it was its input type or output type that was bad.)

Reply:
Yes, we store this information during `get_input_type_hints` which produces the data that's passed to `_add_type_constraint_from_consumer`.  

I added a test called `test_downstream_input_type_hint_error_has_descriptive_error_msg` to confirm this behavior, but let me know if that doesn't cover the case you're talking about.